llvm/include/llvm/Support/Progress.h

```diff
@@ -0,0 +1,256 @@
+//===-- Progress - Generic infrastructure to report progress ----*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_SUPPORT_PROGRESS_H
+#define LLVM_SUPPORT_PROGRESS_H
+
+extern "C" {
+#include <unistd.h>
+}
+
+#include <algorithm>
+#include <chrono>
+#include <optional>
+#include <set>
+#include <vector>
+
+#include "llvm/ADT/DenseMap.h"
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/ADT/SmallPtrSet.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/ADT/Twine.h"
+#include "llvm/Support/ManagedStatic.h"
+#include "llvm/Support/RWMutex.h"
+
+namespace llvm {
+
+class Progress;
+class TaskStack;
+
+/// Track advancement of a task
+///
+/// A task can have a name and optionally a set of maximum steps.
+class Task {
+protected:
+  TaskStack &Stack;
+  int64_t StepIndex = -1;
+  std::optional<int64_t> TotalSteps;
+
+private:
+  std::string TaskName;
+  std::string StepName;
+  bool SubtaskCreated = false;
+  bool Completed = false;
+  bool CurrentStepHasSingleSubtask = false;
+
+public:
+  Task(std::optional<int64_t> TotalSteps, const llvm::Twine &TaskName = "");
+
+  Task(TaskStack &Stack, std::optional<int64_t> TotalSteps,
+       const llvm::Twine &TaskName = "");
```

Comment:
I'd put `TaskStack &Stack` as last argument, so from a user's perspective there are less things to remember: to specify a `TaskStack` just add it as last argument of the constructor.

Also, notice that if we do like I suggested and make `TaskStack`s own `Task`s you wouldn't even need 2 separate constructors anymore.

Of course it would make the construction of a `Task` more verbose, but explicit is better than implicit. And overall I'm not even sure there's a real need make `TaskStack`s customizable. I'd just make the default one the only one. Or are you already planning to use custom `TaskStack`s somewhere. IMHO, customizing them seems a recipe for failure, don't you think?

Reply:
> I'm not even sure there's a real need make TaskStacks customizable

I'm debating this too. If they are not customizable, this feature cannot be reused in isolation. But maybe that's not really necessary.

Reply:
> If they are not customizable, this feature cannot be reused in isolation. 

That's exactly the point I'm questioning. Why would someone want to use the feature in isolation? And specifically, why would someone want to have many different task stacks alive at the same time? Is this additional configurability needed for some expected use case? And if it's not, would it be hard to add it only if and when we really need it?

Reply:
We'd be coupling the progress reporting to the concept of thread, that's all.
One might want to report progress on... a project roadmap! `:)`

There are some scenarios in which execution units are scheduled on a random system thread depending on availability in a thread pool. This is what happens with goroutines.

Not an actual use case, mostly a matter of principle...
Nothing hard to add again, in case.

Reply:
I know, in principle it's always good to decouple. But if you do this on a matter of principle, then the principle has to hold for real in all the parts of the design. And in a way the progress reporting is already coupled to the concept of thread, given that `Progress`, `TaskStack` and `Task` are inextricably tied to each other, and `Progress` uses the thread ID to get the `TaskStack`. So at this point I'd make the whole thing unequivocal and couple progress reporting to threads explicitly. So nobody can have the impression that they are decoupled when they are not, and at least if we really want to decouple the concepts we'll have a practical use case in our hands and we'll think through the implications in all the parts of the design.

Reply:
Agreed.
I thought about trying to get a unique identifier for the current stack as opposed to use the thread ID, but there seems to be no efficient (and portable) way to get, say, the address of the bottom of the stack.

I'll just drop the constructor taking `ProgressReport`. This should enable a couple more simplifications. I can probably drop `Progress::PerThreadStacks` (and its lock) in favor of thread-local global `TaskStack`s, which makes me happy.

Reply:
Done!